sklearn/cluster/_birch.py

```diff
@@ -512,7 +512,32 @@ def fit(self, X, y=None):
         self
             Fitted estimator.
         """
-        # TODO: Remove deprecated flags in 1.2
+
+        check_scalar(
+            self.threshold,
+            "threshold",
+            target_type=numbers.Real,
+            min_val=0.0,
+            include_boundaries="neither",
+        )
+        check_scalar(
+            self.branching_factor,
+            "branching_factor",
+            target_type=numbers.Integral,
+            min_val=1,
+            include_boundaries="neither",
+        )
+        if isinstance(self.n_clusters, numbers.Number):
+            check_scalar(
+                self.n_clusters,
+                "n_clusters",
+                target_type=numbers.Integral,
+                min_val=1,
+                max_val=_num_samples(X),
```

Comment:
On `main`, we currently allow `n_clusters` to be greater than the number of samples:

```python
from sklearn.cluster import Birch
X = [[0, 1], [0.3, 1], [-0.3, 1], [0, -1], [0.3, -1], [-0.3, -1]]
brc = Birch(n_clusters=10)
brc.fit(X)

# Raises warning:
# ConvergenceWarning: Number of subclusters found (2) by BIRCH is less than (10). Decrease the threshold.
```

Restricting `n_clusters` to the number of samples would be BC breaking.

To not block the rest of the PR, I think we can remove `max_val` and keep the other parts of the check:

```suggestion
```